engineapi/backing_image_manager.go

```diff
@@ -0,0 +1,175 @@
+		if err := CheckBackingImageManagerCompatibilty(bim.Status.APIMinVersion, bim.Status.APIVersion); err != nil {
+			return nil, fmt.Errorf("cannot launch a client for incompatible backing image manager %v", bim.Name)
+		}
+	}
+
+	return &BackingImageManagerClient{
+		ip:            bim.Status.IP,
+		apiMinVersion: bim.Status.APIMinVersion,
+		apiVersion:    bim.Status.APIVersion,
+		grpcClient:    bimclient.NewBackingImageManagerClient(fmt.Sprintf("%s:%d", bim.Status.IP, BackingImageManagerDefaultPort)),
+	}, nil
+}
+
+func (c *BackingImageManagerClient) parseBackingImageFileInfo(bi *bimapi.BackingImage) *types.BackingImageFileInfo {
+	if bi == nil {
+		return nil
+	}
+	return &types.BackingImageFileInfo{
+		Name:      bi.Name,
+		URL:       bi.URL,
+		UUID:      bi.UUID,
+		Size:      bi.Size,
+		Directory: bi.Directory,
+
+		State:                types.BackingImageDownloadState(bi.Status.State),
+		Message:              bi.Status.ErrorMsg,
+		SendingReference:     bi.Status.SendingReference,
+		SenderManagerAddress: bi.Status.SenderManagerAddress,
+		DownloadProgress:     bi.Status.DownloadProgress,
+	}
+}
+
+func (c *BackingImageManagerClient) Pull(name, url, uuid string) (*types.BackingImageFileInfo, error) {
+	if err := CheckBackingImageManagerCompatibilty(c.apiMinVersion, c.apiVersion); err != nil {
+		return nil, err
+	}
+	resp, err := c.grpcClient.Pull(name, url, uuid)
+	if err != nil {
+		return nil, err
+	}
+	return c.parseBackingImageFileInfo(resp), nil
+}
+
+func (c *BackingImageManagerClient) Sync(name, url, uuid, fromHost, toHost string, size int64) (*types.BackingImageFileInfo, error) {
+	if err := CheckBackingImageManagerCompatibilty(c.apiMinVersion, c.apiVersion); err != nil {
+		return nil, err
+	}
+	resp, err := c.grpcClient.Sync(name, url, uuid, fromHost, toHost, size)
+	if err != nil {
+		return nil, err
+	}
+	return c.parseBackingImageFileInfo(resp), nil
+}
+
+func (c *BackingImageManagerClient) Delete(name string) error {
+	if err := CheckBackingImageManagerCompatibilty(c.apiMinVersion, c.apiVersion); err != nil {
+		return err
+	}
+	return c.grpcClient.Delete(name)
+}
+
+func (c *BackingImageManagerClient) Get(name string) (*types.BackingImageFileInfo, error) {
+	if err := CheckBackingImageManagerCompatibilty(c.apiMinVersion, c.apiVersion); err != nil {
+		return nil, err
+	}
+	backingImage, err := c.grpcClient.Get(name)
+	if err != nil {
+		return nil, err
+	}
+	return c.parseBackingImageFileInfo(backingImage), nil
+}
+
+func (c *BackingImageManagerClient) List() (map[string]types.BackingImageFileInfo, error) {
+	if err := CheckBackingImageManagerCompatibilty(c.apiMinVersion, c.apiVersion); err != nil {
+		return nil, err
+	}
+	backingImages, err := c.grpcClient.List()
+	if err != nil {
+		return nil, err
+	}
+	result := map[string]types.BackingImageFileInfo{}
+	for name, backingImage := range backingImages {
+		result[name] = *c.parseBackingImageFileInfo(backingImage)
```

Comment:
Do you mean `client.List(ctx, &empty.Empty{})` should return `map[string]api.BackingImage` rather than `map[string]*api.BackingImage`?

In this function, there won't be a nil as the value of the returned map.

Reply:
Yes in reality, it will not :) 

But from the code, theoretically it's possible because parseBackingImageFileinfo will possibly return nil.

Let's keep it for now!